.travis.yml

```diff
@@ -37,13 +37,19 @@ env:
 stages:
   - flake8
   - test
+  - docker-push
 
 jobs:
   include:
     - stage: flake8
       name: Flake8
       script: docker-compose --log-level ERROR -f scripts/ci/docker-compose.yml run airflow-testing /app/scripts/ci/run-ci.sh
       env: TOX_ENV=flake8
+    - name: docker-push
+      # require the branch name to be master (note for PRs this is the base branch name)
+      if: branch = master
+      script: docker build -t apache/airflow:master -t apache/airflow:$TRAVIS_COMMIT . && docker push apache/airflow:$TRAVIS_COMMIT && docker push apache/airflow:master
```

Comment:
You will have to use `docker pull` followed by `docker build --cache-from` in order to utilise docker caching. The problem is that when you start builds on a fresh machine the layers are different from what was built on another machine and effectively you do not reuse previously built layers - i.e. the image is re-build and all layers are pushed every time you make push.

There is relatively new --cache-from feature of docker which works in the way that when you pull the image built elsewhere you can use it as source of cache for subsequent builds. This means that you will only incrementally built final layers  only if for example just sources or just setup.py changes.

You can read more about it in this issue : https://stackoverflow.com/questions/37440705/how-to-setup-docker-to-use-cache-from-registry-on-every-build-step (see the second answer - highest voted)

Reply:
Another option - we could consider using Kaniko. Kaniko is a great tool provided by Google to build reproducible images that you can use inside another Docker container as well. It has much better support for building reproducible images (independent from the machine it is built on):

https://github.com/GoogleContainerTools/kaniko

Reply:
You do realise that this part of the code is already gone? The building will happen by docker hub itself once this is merged.

Reply:
Ah. I see now. I was looking at commits rather than final change. Indeed in this case Dockerhub manages caching on its own. All good then. 👍

Reply:
BTW. Not sure if this is the right moment, but maybe we could incorporate some of the layerin and concepts from what we've implemented in our GCP-operator related Airflow Development environment (We call it airflow-breeze). It has a number of things implemented such as nicer layering, ability to only rebuild different layers separately rather than only airflow etc. It's much more cache friendly and I think some of the concepts from it might be incorporated in the official image as well: 

https://github.com/PolideaInternal/airflow-breeze/blob/master/Dockerfile

Reply:
Thanks @potiuk for the feedback. I think this Dockerfile is fine for a first step, after that we can improve and make it nicer. The idea is to have a build each time there is a commit on master, not on the other branches.
Right now this is done by the automatic buiding tool of Dockerhub: https://jira.apache.org/jira/browse/INFRA-17595
There is no way to use caching since it will kick off a fresh build every time.

Feel free to improve the image, highly appreciated.

Reply:
I wasn't aware of that, thanks for pointing out.